### PR TITLE
NAS-105595 / 12.0 / Retrieve correct primary key on insertion of record in database

### PR DESCRIPTION
**Pull request by sonicaj**
This PR adds changes to ensure that we retrieve correct primary key on insertion of a record in database.

### CODE REVIEW

**Comment by sonicaj** [comment]
So that it returns the complete row it inserted instead of just the id ?
We use `datastore.insert` everywhere currently to only retrieve the id on `do_create`. We could get the `id` from the row returned. Do you think we should proceed with changing this behaviour ?

**Comment by sonicaj** [comment]
Also for this specific case, `disk_identifier` was already set. Generally it works well to only get the `id` from `datastore.insert` - alternate would be something like `data['id'] = middleware_call('datastore.insert', data)['id']`

**Comment by themylogin** [comment]
It should return primary key, we are not interested in `last_insert_rowid()` if we are not using rowid as primary key. `_handle_relationships` will also fail with this approach.

File: src/middlewared/middlewared/plugins/datastore/write.py [hunk 1]
```diff
@@ -29,7 +28,11 @@ class Config:
                     insert.setdefault(column.name, '')
 
         await self.middleware.call('datastore.execute_write', table.insert().values(**insert))
-        pk = (await self.middleware.call('datastore.fetchall', 'SELECT last_insert_rowid()'))[0][0]
+        row_id = (await self.middleware.call('datastore.fetchall', 'SELECT last_insert_rowid()'))[0][0]
+        pk = (await self.middleware.call(
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/datastore/write.py:32]: We can probably avoid extra query. If primary key is integer, we return rowid. If primary key is not integer, we already have it in `insert`. Also, please add tests.

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/datastore/write.py:32]: That would be generally true ( at least right now it is I think ) but if say we were to have a table which has an integer as a primary key which is not `id` ?

Comment (themylogin) [reply src/middlewared/middlewared/plugins/datastore/write.py:32]: In SQLite integer primary key will automatically be an alias for rowid

File: src/middlewared/middlewared/plugins/datastore/write.py [hunk 2]
```diff
@@ -29,7 +28,11 @@ class Config:
                     insert.setdefault(column.name, '')
 
         await self.middleware.call('datastore.execute_write', table.insert().values(**insert))
-        pk = (await self.middleware.call('datastore.fetchall', 'SELECT last_insert_rowid()'))[0][0]
+        pk_column = self._get_pk(table)
+        if isinstance(pk_column.type, (types.String, types.TEXT)):
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/datastore/write.py:32]: The only type for primary key being rowid is integer, for all other types (including e.g. bigint sqlite requires us to specify pk explicitly)

File: src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py [hunk 3]
```diff
@@ -636,6 +639,44 @@ class DefaultModel(Model):
...
+
+        assert len(await ds.query("test.stringprimarykey", [["string_id", "=", pk]])) == 1
+
+
+@pytest.mark.asyncio
+async def test__insert_default_integer_pk_value():
+    async with datastore_test() as ds:
+        pk = await ds.insert("test.default", {"string": "VALUE"})
+
+        assert len(await ds.query("test.default", [["id", "=", pk]])) == 1
```
Comment (themylogin) [inline src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py:669]: What does this test check for?

File: src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py [hunk 4]
```diff
@@ -636,6 +639,44 @@ class DefaultModel(Model):
...
+    integer_id = sa.Column(sa.String(100), primary_key=True)
+    value = sa.Column(sa.Integer(), nullable=True)
+
+
+@pytest.mark.asyncio
+async def test__insert_string_pk_record():
+    async with datastore_test() as ds:
+        pk = await ds.insert("test.stringprimarykey", {"string_id": "unique_key", "value": 1})
+
+        assert len(await ds.query("test.stringprimarykey", [["string_id", "=", pk]])) == 1
```
Comment (themylogin) [inline src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py:661]: Won't it be better to assert returned value from ds.insert?

Comment (sonicaj) [reply src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py:661]: I think this is better as we ensure that the value returned is actually the primary key as reflected by the database and not somehow just being returned from `ds.insert`

File: src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py (hunk 3 above)
Comment (sonicaj) [reply src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py:669]: So it checks to ensure the case where we have integer as the primary key

File: src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py (hunk 4 above)
Comment (themylogin) [reply src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py:661]: We can do both asserts

Comment (themylogin) [reply src/middlewared/middlewared/pytest/unit/plugins/test_datastore.py:661]: Tests like this tend to hide situations where two wrongs make one right (but are still wrongs)

**Pull request merged**